adapter_pipelines/cellranger/adapter.wdl

```diff
@@ -0,0 +1,268 @@
+          inputs.append({"name": "expect_cells", "value": "${expect_cells}"})
+
+      print('write inputs.tsv')
+      with open('inputs.tsv', 'w') as f:
+          f.write('name\tvalue\n')
+          for input in inputs:
+              print(input)
+              f.write('{0}\t{1}\n'.format(input['name'], input['value']))
+      print('finished')
+      CODE
+    >>>
+
+    runtime {
+      docker: "quay.io/humancellatlas/secondary-analysis-pipeline-tools:" + pipeline_tools_version
+    }
+
+    output {
+      File inputs = "inputs.tsv"
+    }
+}
+
+workflow Adapter10xCount {
+  String bundle_uuid
+  String bundle_version
+
+  String reference_name
+  File transcriptome_tar_gz
+  Int? expect_cells
+
+  # Submission
+  File format_map
+  String dss_url
+  String submit_url
+  String method
+  String schema_url
+  String cromwell_url
+  String analysis_process_schema_version
+  String analysis_protocol_schema_version
+  String analysis_file_version
+  String run_type
+  Int? retry_max_interval
+  Float? retry_multiplier
+  Int? retry_timeout
+  Int? individual_request_timeout
+  String reference_bundle
+  Boolean use_caas
+
+  # Set runtime environment such as "dev" or "staging" or "prod" so submit task could choose proper docker image to use
+  String runtime_environment
+  # By default, don't record http requests, unless we override in inputs json
+  Boolean record_http = false
+  Int max_cromwell_retries = 0
+  Boolean add_md5s = false
+
+  String pipeline_tools_version = "se-10x-adapter-wdl"
+
+  call GetInputs {
+    input:
+      bundle_uuid = bundle_uuid,
+      bundle_version = bundle_version,
+      dss_url = dss_url,
+      retry_multiplier = retry_multiplier,
+      retry_max_interval = retry_max_interval,
+      retry_timeout = retry_timeout,
+      individual_request_timeout = individual_request_timeout,
+      record_http = record_http,
+      pipeline_tools_version = pipeline_tools_version
+  }
+
+  # Cellranger code in 10x count wdl requires files to be named a certain way.
+  # To accommodate that, rename_files copies the blue box files into the
+  # cromwell execution bucket but with the names cellranger expects.
+  # Putting this in its own task lets us take advantage of automatic localizing
+  # and delocalizing by Cromwell/JES to actually read and write stuff in buckets.
+  # TODO: Replace scatter with a for-loop inside of the task to avoid creating a
+  # VM for each set of files that needs to be renamed
+  scatter(i in range(length(GetInputs.lanes))) {
+    call rename_files as prep {
+      input:
+        r1 = GetInputs.r1_fastq[i],
+        r2 = GetInputs.r2_fastq[i],
+        i1 = GetInputs.i1_fastq[i],
+        sample_id = GetInputs.sample_id,
+        lane = GetInputs.lanes[i],
+        pipeline_tools_version = pipeline_tools_version
+      }
+    }
+
+  call CellRanger.CellRanger as analysis {
+    input:
+      sample_id = GetInputs.sample_id,
+      fastqs = flatten([prep.r1_new, prep.r2_new, prep.i1_new]),
```

Comment:
So if you have two lanes, will this order like [L1_R1, L2_R1, L1_R2, L2_R2, L1_I1, L2_I1]? I was surprised it's not ordered by lane instead: [L1_R1, L1_R2, L1_I1, L2_R1, L2_R2, L2_I1]. But if that's what cellranger wants then that's what it wants.

Reply:
The order here doesn't matter because the cellranger.wdl is getting the directories containing the fastq files from the file paths: https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/cellranger/cellranger.wdl#L88

Reply:
Ah, I get it now. Might be worth adding a code comment for posterity.